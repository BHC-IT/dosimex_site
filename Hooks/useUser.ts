import { useState, useEffect } from 'react';

export interface IUser {
	username: string,
	password: string,
}


const useUser = () => {

	const [token, setToken] = useState<string | null>(null)

	useEffect(() => {
		setToken(localStorage.getItem('access_token'))
	}, []);

	return token;
}

export default useUser;
