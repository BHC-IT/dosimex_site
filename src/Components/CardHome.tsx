import * as React from 'react';
import Link from 'next/link';
import Image from 'next/image';
import Radium from 'radium';
import * as CSS from 'csstype';

export interface IStyles {
	button: CSS.Properties,
	subtitle: CSS.Properties,
	card: CSS.Properties,
	image: CSS.Properties,
}

interface IProps {
	icon: string,
	title: string,
	content: string,
	route: string,
}

const CardHome = (props: IProps) => {

	return (
		<div style={styles.card}>
			<div style={styles.image}>
				<Image
					src={props.icon}
					alt="icône"
					width={158}
					height={158}
				/>
			</div>
			<h4 style={{marginBottom: "0"}}>{props.title}</h4>
			<p style={styles.subtitle}>{props.content}</p>
			<button style={styles.button}>
				<Link href={`/${props.route}`} replace><span>→</span></Link>
			</button>
		</div>
	);

}

export default Radium(CardHome);

export const styles =  {
	button: {
		width: "40px",
		height: "40px",
		display: "flex",
		alignItems: "center",
		justifyContent: "center",
		fontSize: "3rem",
		padding: "20px",
		marginLeft: "auto",
		marginRight: "auto",
		marginTop: "2vh",
		marginBottom: "4vh",
		backgroundColor: "var(--main)",
		borderRadius: "50%",
		color: "white",
		textTransform: "uppercase",
		transition: "all 0.3s ease 0s",
		':hover': {
			color: "white",
			transform: "translateY(-2px)",
			boxShadow: "0px 5px 5px rgba(0, 0, 0, 0.1)",
		},
	},
	card: {
		borderRadius: "10px",
		boxShadow: "0px 5px 5px rgba(0, 0, 0, 0.1)",
		textAlign: "center",
		width: "28%",
	},
	subtitle: {
		color: "var(--grey)",
		fontSize: "1.8rem",
		marginBottom: "4vh",
		marginLeft: "2%",
		marginRight: "2%",
	},
	image: {
		width: "20%",
		height: "20%",
		marginLeft: "auto",
		marginRight: "auto",
	},
}