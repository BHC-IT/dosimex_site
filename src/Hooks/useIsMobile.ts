import { useState, useEffect } from 'react'

const TIMEOUT_DELAY = 10

// Mobile breakpoint - commonly used in responsive design
const MOBILE_BREAKPOINT = 768
const TABLET_BREAKPOINT = 1024

// Helper function to detect mobile based on screen width and user agent
const detectIsMobile = (): boolean => {
	if (typeof window === 'undefined') return false

	// Primary detection via screen width (works with responsive design mode)
	const screenWidth = window.innerWidth
	if (screenWidth < MOBILE_BREAKPOINT) return true

	// Secondary detection via user agent (for actual mobile devices)
	const userAgent = navigator.userAgent.toLowerCase()
	const mobileKeywords = [
		'android',
		'webos',
		'iphone',
		'ipad',
		'ipod',
		'blackberry',
		'windows phone',
		'mobile',
	]

	return mobileKeywords.some(keyword => userAgent.includes(keyword))
}

// Helper function to detect tablet based on screen width and user agent
const detectIsTablet = (): boolean => {
	if (typeof window === 'undefined') return false

	const screenWidth = window.innerWidth

	// Tablet range: between mobile and desktop breakpoints
	if (screenWidth >= MOBILE_BREAKPOINT && screenWidth < TABLET_BREAKPOINT) {
		return true
	}

	// Additional tablet detection via user agent
	const userAgent = navigator.userAgent.toLowerCase()
	const tabletKeywords = ['ipad', 'tablet', 'kindle']

	return tabletKeywords.some(keyword => userAgent.includes(keyword))
}

export const useIsMobile = <T>(styles: (m: boolean) => T): T | null => {
	const [style, setStyle] = useState<T | null>(null)

	useEffect(() => {
		const updateStyle = () => {
			const isMobile = detectIsMobile()
			setStyle(styles(isMobile))
		}

		// Initial calculation
		const timeoutId = setTimeout(updateStyle, TIMEOUT_DELAY)

		// Listen for window resize to handle responsive design mode
		const handleResize = () => {
			updateStyle()
		}

		window.addEventListener('resize', handleResize)

		return () => {
			clearTimeout(timeoutId)
			window.removeEventListener('resize', handleResize)
		}
	}, [styles])

	return style
}

export const useMobile = (): boolean => {
	const [mobile, setMobile] = useState(false) // Default to false for SSR/initial render

	useEffect(() => {
		const updateMobile = () => {
			setMobile(detectIsMobile())
		}

		// Initial calculation
		const timeoutId = setTimeout(updateMobile, TIMEOUT_DELAY)

		// Listen for window resize to handle responsive design mode
		const handleResize = () => {
			updateMobile()
		}

		window.addEventListener('resize', handleResize)

		return () => {
			clearTimeout(timeoutId)
			window.removeEventListener('resize', handleResize)
		}
	}, [])

	return mobile
}

export const useTablet = (): boolean => {
	const [tablet, setTablet] = useState(false) // Default to false for SSR/initial render

	useEffect(() => {
		const updateTablet = () => {
			setTablet(detectIsTablet())
		}

		// Initial calculation
		const timeoutId = setTimeout(updateTablet, TIMEOUT_DELAY)

		// Listen for window resize to handle responsive design mode
		const handleResize = () => {
			updateTablet()
		}

		window.addEventListener('resize', handleResize)

		return () => {
			clearTimeout(timeoutId)
			window.removeEventListener('resize', handleResize)
		}
	}, [])

	return tablet
}

export const useDeviceType = (): { isMobile: boolean; isTablet: boolean } => {
	const mobile = useMobile()
	const tablet = useTablet()

	return { isMobile: mobile, isTablet: tablet }
}
