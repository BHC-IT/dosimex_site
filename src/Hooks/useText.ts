import { useRouter } from 'next/router';

import ILang from '../lang/interface';

import fr from '../lang/fr';
import en from '../lang/en';
import debug from '../lang/debug';

export const text : {[$:string]: ILang} = {fr, en, debug};

export const useText = (page: string) : any => {
	const route = useRouter();

	if (route.locale === 'debug')
		return text[route.locale][page];
	else if (route?.locale)
		return text[route.locale.slice(0,2)]?.[page];
	return null
}
