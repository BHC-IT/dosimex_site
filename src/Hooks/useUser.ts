import { useState, useEffect } from 'react';


export interface IUser {
	username: string,
	password: string,
}

const remove = (setToken: Function) => {
	localStorage.removeItem('access_token')
	localStorage.removeItem('exp_token')
	setToken(null)
}

const useUser = () => {

	const [token, setToken] = useState<string | null>(null)

	useEffect(() => {
		setToken(localStorage.getItem('access_token'))
		const exp = Number(localStorage.getItem('exp_token'))

		if (!exp) {
			remove(setToken)
			return;
		}
		if (exp < Date.now())
			remove(setToken)
		else
			setTimeout(() => remove(setToken), exp - Date.now())
	}, []);

	return token;
}

export default useUser;
