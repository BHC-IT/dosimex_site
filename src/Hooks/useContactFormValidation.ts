import emailjs from '@emailjs/browser'
import { useState, useCallback } from 'react'
import { isPossiblePhoneNumber } from 'react-phone-number-input'
import { toast } from 'react-toastify'

import { IValidator } from '../Components/Input'

// Email validation regex
const EMAIL_REGEX =
	/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

// Validation functions
export const isInputValid = (value: string): boolean => value.trim() !== ''
export const isEmailValid = (value: string): boolean => EMAIL_REGEX.test(value)
export const isPhoneValidNumber = (phone: string | undefined): boolean =>
	!phone || isPossiblePhoneNumber(phone)

// Validator factory functions
export const createRequiredValidator = (errorMessage: string): IValidator => ({
	validationFunction: value => isInputValid(value),
	errorMessage,
})

export const createEmailValidator = (errorMessage: string): IValidator => ({
	validationFunction: value => isEmailValid(value),
	errorMessage,
})

export interface IContactFormValidation {
	// Form field states
	name: string | null
	email: string | null
	phone: string | undefined
	enterprise: string | null
	address: string | null
	subject: string | null
	message: string | null

	// Validation states
	nameValid: boolean
	phoneError: string | null
	isLoading: boolean

	// Form handlers
	setName: (value: string) => void
	setEmail: (value: string) => void
	setPhone: (value: string | undefined) => void
	setEnterprise: (value: string) => void
	setAddress: (value: string) => void
	setSubject: (value: string) => void
	setMessage: (value: string) => void
	setNameValid: (valid: boolean) => void
	setIsLoading: (loading: boolean) => void

	// Validation methods
	validatePhone: () => void
	validateForm: () => boolean
	resetForm: () => void
	handleSubmit: (e: React.FormEvent) => void

	// Validator configurations
	getNameValidators: (errorMessage: string) => IValidator[]
	getEmailValidators: (errorEmail: string[]) => IValidator[]
	getAddressValidators: (errorMessage: string) => IValidator[]
}

// Validation helper functions
const validateName = (name: string | null): boolean => {
	return name !== null && name.trim() !== ''
}

const validateEmail = (email: string | null): boolean => {
	if (!email || email.trim() === '') return false
	const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
	return emailRegex.test(email)
}

const validateMessage = (message: string | null): boolean => {
	return message !== null && message.trim() !== ''
}

// Custom hook for form state management
const useFormState = () => {
	const [name, setName] = useState<string | null>(null)
	const [email, setEmail] = useState<string | null>(null)
	const [phone, setPhone] = useState<string | undefined>(undefined)
	const [enterprise, setEnterprise] = useState<string | null>(null)
	const [address, setAddress] = useState<string | null>(null)
	const [subject, setSubject] = useState<string | null>(null)
	const [message, setMessage] = useState<string | null>(null)
	const [nameValid, setNameValid] = useState(false)
	const [isLoading, setIsLoading] = useState(false)

	return {
		name, setName,
		email, setEmail,
		phone, setPhone,
		enterprise, setEnterprise,
		address, setAddress,
		subject, setSubject,
		message, setMessage,
		nameValid, setNameValid,
		isLoading, setIsLoading,
	}
}

// Form validation logic
const useFormValidation = (nameValid: boolean, phone: string | undefined) => {
	const checkInputs = useCallback((): boolean => {
		return nameValid && (phone ? isPossiblePhoneNumber(phone) : true)
	}, [nameValid, phone])

	const validateForm = useCallback(() => checkInputs(), [checkInputs])

	const validatePhone = useCallback(() => {
		// Phone validation is done in checkInputs
	}, [])

	return { checkInputs, validateForm, validatePhone }
}

// Form reset functionality
const useFormReset = (
	setName: (value: string | null) => void,
	setEmail: (value: string | null) => void,
	setPhone: (value: string | undefined) => void,
	setEnterprise: (value: string | null) => void,
	setAddress: (value: string | null) => void,
	setSubject: (value: string | null) => void,
	setMessage: (value: string | null) => void,
	setNameValid: (valid: boolean) => void,
	setIsLoading: (loading: boolean) => void,
) => {
	const resetForm = useCallback(() => {
		setName(null)
		setEmail(null)
		setPhone(undefined)
		setEnterprise(null)
		setAddress(null)
		setSubject(null)
		setMessage(null)
		setNameValid(false)
		setIsLoading(false)
	}, [setName, setEmail, setPhone, setEnterprise, setAddress, setSubject, setMessage, setNameValid, setIsLoading])

	return { resetForm }
}

// Email sending functionality
const useEmailSending = (
	name: string | null,
	email: string | null,
	phone: string | undefined,
	enterprise: string | null,
	address: string | null,
	subject: string | null,
	message: string | null,
	text: Record<string, string | string[]> | undefined,
	setIsLoading: (loading: boolean) => void,
	resetForm: () => void,
) => {
	const sendEmail = useCallback((): void => {
		sendContactEmail({
			name, email, phone, enterprise, address, subject, message,
			text, setIsLoading, resetForm,
		})
	}, [name, email, phone, enterprise, address, subject, message, text, setIsLoading, resetForm])

	return { sendEmail }
}

// Validator creation functions
const useValidatorCreation = (
	name: string | null,
	email: string | null,
	message: string | null,
) => {
	const getNameValidators = useCallback((errorName: string): IValidator[] => [
		{
			validationFunction: () => validateName(name),
			errorMessage: errorName || 'Name is required',
		},
	], [name])

	const getEmailValidators = useCallback((errorEmail: string[]): IValidator[] => [
		{
			validationFunction: () => email !== null && email.trim() !== '',
			errorMessage: errorEmail[0] || 'Email is required',
		},
		{
			validationFunction: () => validateEmail(email),
			errorMessage: errorEmail[1] || 'Email format is invalid',
		},
	], [email])

	const getAddressValidators = useCallback((errorMessage: string): IValidator[] => [
		{
			validationFunction: () => validateMessage(message),
			errorMessage: errorMessage || 'Message is required',
		},
	], [message])

	return { getNameValidators, getEmailValidators, getAddressValidators }
}

// Form submission handling
const useFormSubmission = (checkInputs: () => boolean, sendEmail: () => void) => {
	const handleSubmit = useCallback((e: React.FormEvent) => {
		e.preventDefault()
		if (checkInputs()) {
			sendEmail()
		}
	}, [checkInputs, sendEmail])

	return { handleSubmit }
}

export const useContactFormValidation = (text?: Record<string, string | string[]>): IContactFormValidation => {
	const formState = useFormState()
	const formValidation = useFormValidation(formState.nameValid, formState.phone)
	const { resetForm } = useFormReset(
		formState.setName,
		formState.setEmail,
		formState.setPhone,
		formState.setEnterprise,
		formState.setAddress,
		formState.setSubject,
		formState.setMessage,
		formState.setNameValid,
		formState.setIsLoading,
	)
	const { sendEmail } = useEmailSending(
		formState.name,
		formState.email,
		formState.phone,
		formState.enterprise,
		formState.address,
		formState.subject,
		formState.message,
		text,
		formState.setIsLoading,
		resetForm,
	)
	const validatorCreation = useValidatorCreation(formState.name, formState.email, formState.message)
	const { handleSubmit } = useFormSubmission(formValidation.checkInputs, sendEmail)

	return {
		// Form field states
		name: formState.name,
		email: formState.email,
		phone: formState.phone,
		enterprise: formState.enterprise,
		address: formState.address,
		subject: formState.subject,
		message: formState.message,

		// Validation states
		nameValid: formState.nameValid,
		phoneError: null, // No longer tracking phone errors separately
		isLoading: formState.isLoading,

		// Form handlers
		setName: formState.setName,
		setEmail: formState.setEmail,
		setPhone: formState.setPhone,
		setEnterprise: formState.setEnterprise,
		setAddress: formState.setAddress,
		setSubject: formState.setSubject,
		setMessage: formState.setMessage,
		setNameValid: formState.setNameValid,
		setIsLoading: formState.setIsLoading,

		// Validation methods
		validatePhone: formValidation.validatePhone,
		validateForm: formValidation.validateForm,
		resetForm,
		handleSubmit,

		// Validator configurations
		getNameValidators: validatorCreation.getNameValidators,
		getEmailValidators: validatorCreation.getEmailValidators,
		getAddressValidators: validatorCreation.getAddressValidators,
	}
}

// Email sending helper function
const sendContactEmail = (params: {
	name: string | null, email: string | null, phone: string | undefined,
	enterprise: string | null, address: string | null, subject: string | null,
	message: string | null, text: Record<string, string | string[]> | undefined,
	setIsLoading: (loading: boolean) => void,
	resetForm: () => void
}) => {
	const { name, email, phone, enterprise, address, subject, message, text, setIsLoading, resetForm } = params
	const toastLoad = toast.dark(text?.sending)
	setIsLoading(true)
	emailjs
		.send(
			process.env.NEXT_PUBLIC_EMAILJS_SERVICE_ID ?? '',
			process.env.NEXT_PUBLIC_EMAILJS_TEMPLATE_ID ?? '',
			{
					text: message,
					name,
					email,
					phone,
					enterprise,
					address,
					subject,
				},
				process.env.NEXT_PUBLIC_EMAILJS_PUBLIC_KEY ?? '',
			)
			.then(() => {
				toast.dismiss(toastLoad)
				toast.success(text?.messageSent)
				resetForm()
			})
			.catch(() => {
				toast.dismiss(toastLoad)
				toast.error(text?.messageNotSent)
			})
			.finally(() => {
				setIsLoading(false)
			})
}
