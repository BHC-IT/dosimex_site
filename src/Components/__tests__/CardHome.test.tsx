import { render, screen, fireEvent } from '@testing-library/react'
import { describe, it, expect, vi, beforeEach } from 'vitest'

import { useIsMobile } from '../../Hooks/useIsMobile'
import { useText } from '../../Hooks/useText'
import CardHome, { styles } from '../CardHome'

// Mock Next.js router
const mockPush = vi.fn()
vi.mock('next/router', () => ({
	useRouter: vi.fn(() => ({
		push: mockPush,
		pathname: '/',
		query: {},
		asPath: '/',
	})),
}))

// Mock useIsMobile hook
vi.mock('../../Hooks/useIsMobile', () => ({
	useIsMobile: vi.fn(),
}))

// Mock useText hook
vi.mock('../../Hooks/useText', () => ({
	useText: vi.fn(),
}))

// Mock react-parallax-tilt
vi.mock('react-parallax-tilt', () => ({
	default: ({ children, ...props }: any) => (
		<div
			data-testid='tilt-wrapper'
			{...props}
		>
			{children}
		</div>
	),
}))

// Test data factories
const createMockCardProps = (overrides = {}) => ({
	icon: '/Images/software-icon.png',
	title: 'Software Solutions',
	content:
		'Advanced radiation dosimetry software for comprehensive dose management and analysis.',
	route: '/software',
	...overrides,
})

describe('CardHome Component', () => {
	beforeEach(() => {
		vi.clearAllMocks()
		// Always return valid styles, never null
		vi.mocked(useIsMobile).mockReturnValue(styles(false))
		vi.mocked(useText).mockReturnValue({ cardIcon: 'Feature icon' })
	})

	describe('Rendering', () => {
		it('should render card with all content correctly', () => {
			const props = createMockCardProps()
			const { container } = render(<CardHome {...props} />)
			expect(container.firstChild).toMatchSnapshot()
		})

		it('should render component structure correctly', () => {
			const props = createMockCardProps()
			const { container } = render(<CardHome {...props} />)

			// Component should render without errors
			expect(container).toBeDefined()
			expect(container.firstChild).not.toBeNull()
		})

		it('should return null when styles are null (SSR protection)', () => {
			vi.mocked(useIsMobile).mockReturnValue(null)

			const props = createMockCardProps()
			const { container } = render(<CardHome {...props} />)

			expect(container.firstChild).toBeNull()
		})
	})

	describe('Content Display', () => {
		it('should display card title correctly', () => {
			const props = createMockCardProps({
				title: 'Training Programs',
			})

			render(<CardHome {...props} />)

			expect(screen.getByText('Training Programs')).toBeInTheDocument()
		})

		it('should display card content correctly', () => {
			const props = createMockCardProps({
				content: 'Comprehensive training materials and certification programs.',
			})

			render(<CardHome {...props} />)

			expect(
				screen.getByText('Comprehensive training materials and certification programs.'),
			).toBeInTheDocument()
		})

		it('should display icon with proper alt text', () => {
			const props = createMockCardProps({
				icon: '/Images/training-icon.png',
				title: 'Training',
			})

			render(<CardHome {...props} />)

			const iconImage = screen.getByAltText('Feature icon - Training')
			expect(iconImage).toBeInTheDocument()
			// Next.js Image component transforms src URLs
			expect(iconImage.getAttribute('src')).toContain('training-icon.png')
		})

		it('should handle missing alt text gracefully', () => {
			vi.mocked(useText).mockReturnValue(null)

			const props = createMockCardProps({
				title: 'Software',
			})

			render(<CardHome {...props} />)

			const iconImage = screen.getByAltText('Icon - Software')
			expect(iconImage).toBeInTheDocument()
		})
	})

	describe('Navigation Functionality', () => {
		it('should navigate to correct route when clicked', () => {
			const props = createMockCardProps({
				route: '/training',
			})

			render(<CardHome {...props} />)

			const cardButton = screen.getByRole('button')
			fireEvent.click(cardButton)

			expect(mockPush).toHaveBeenCalledWith('/training')
		})

		it('should handle different route formats', () => {
			const routeTestCases = ['/software', '/training', '/manuals', '/', '/about-us']

			routeTestCases.forEach(route => {
				vi.clearAllMocks()
				const props = createMockCardProps({ route })
				const { unmount } = render(<CardHome {...props} />)

				const cardButton = screen.getByRole('button')
				fireEvent.click(cardButton)

				expect(mockPush).toHaveBeenCalledWith(route)
				unmount()
			})
		})

		it('should be accessible with proper ARIA label', () => {
			const props = createMockCardProps({
				title: 'Software Solutions',
				content: 'Advanced dosimetry software',
			})

			render(<CardHome {...props} />)

			const cardButton = screen.getByRole('button')
			expect(cardButton).toHaveAttribute(
				'aria-label',
				'Navigate to Software Solutions section - Advanced dosimetry software',
			)
		})

		it('should be keyboard accessible', () => {
			const props = createMockCardProps()

			render(<CardHome {...props} />)

			const cardButton = screen.getByRole('button')

			// Should be focusable
			cardButton.focus()
			expect(document.activeElement).toBe(cardButton)

			// Note: fireEvent.keyDown doesn't trigger onClick for buttons
			// This is expected behavior, only click events trigger navigation
		})
	})

	describe('Responsive Behavior', () => {
		it('should use mobile styles when mobile is true', () => {
			vi.mocked(useIsMobile).mockReturnValue(styles(true))

			const props = createMockCardProps()
			render(<CardHome {...props} />)

			// Component should render with mobile-specific behavior
			expect(screen.getByRole('button')).toBeInTheDocument()
		})

		it('should use desktop styles when mobile is false', () => {
			vi.mocked(useIsMobile).mockReturnValue(styles(false))

			const props = createMockCardProps()
			render(<CardHome {...props} />)

			// Component should render with desktop-specific behavior
			expect(screen.getByRole('button')).toBeInTheDocument()
		})
	})

	describe('Tilt Effect Integration', () => {
		it('should wrap content in Tilt component', () => {
			const props = createMockCardProps()
			render(<CardHome {...props} />)

			const tiltWrapper = screen.getByTestId('tilt-wrapper')
			expect(tiltWrapper).toBeInTheDocument()
		})
	})

	describe('Visual Elements', () => {
		it('should display FontAwesome arrow icon', () => {
			const props = createMockCardProps()
			render(<CardHome {...props} />)

			// Check if FontAwesome icon container exists
			const button = screen.getByRole('button')
			const arrowContainer =
				button.querySelector('[data-icon]') ?? button.querySelector('svg')
			expect(arrowContainer).toBeTruthy()
		})

		it('should have proper button styling structure', () => {
			const props = createMockCardProps()
			render(<CardHome {...props} />)

			const cardButton = screen.getByRole('button')
			// Check for the inline styles set on the button
			expect(cardButton.style.background).toBe('white')
			// Note: border style gets normalized to empty string in test environment
			expect(cardButton.style.border || 'none').toBe('none')
		})
	})

	describe('Edge Cases', () => {
		it('should handle empty string props gracefully', () => {
			const props = createMockCardProps({
				title: '',
				content: '',
				icon: '',
				route: '',
			})

			const { container } = render(<CardHome {...props} />)

			// Should still render component structure
			expect(container.firstChild).not.toBeNull()
			expect(screen.getByRole('button')).toBeInTheDocument()
		})

		it('should handle special characters in text', () => {
			const props = createMockCardProps({
				title: 'Logiciels & Solutions',
				content: 'Solutions avancées pour la dosimétrie : mesures précises à 99,9%!',
			})

			render(<CardHome {...props} />)

			expect(screen.getByText('Logiciels & Solutions')).toBeInTheDocument()
			expect(screen.getByText(/Solutions avancées pour la dosimétrie/)).toBeInTheDocument()
		})

		it('should handle long text content', () => {
			const props = createMockCardProps({
				title: 'Advanced Radiation Dosimetry Software Solutions',
				content:
					'Comprehensive radiation dose management and analysis software featuring real-time monitoring capabilities, advanced reporting tools, regulatory compliance features, and user-friendly interface designed for healthcare professionals and researchers.',
			})

			render(<CardHome {...props} />)

			expect(
				screen.getByText('Advanced Radiation Dosimetry Software Solutions'),
			).toBeInTheDocument()
			expect(screen.getByText(/Comprehensive radiation dose management/)).toBeInTheDocument()
		})

		it('should handle navigation errors gracefully', () => {
			mockPush.mockRejectedValue(new Error('Navigation failed'))

			const props = createMockCardProps()
			render(<CardHome {...props} />)

			const cardButton = screen.getByRole('button')

			// Should not throw error when navigation fails
			expect(() => {
				fireEvent.click(cardButton)
			}).not.toThrow()
		})
	})

	describe('Image Optimization', () => {
		it('should use proper image optimization settings', () => {
			const props = createMockCardProps()
			render(<CardHome {...props} />)

			const image = screen.getByRole('img')
			expect(image).toHaveAttribute('width', '80')
			expect(image).toHaveAttribute('height', '80')
			expect(image).toHaveAttribute('loading', 'lazy')
		})

		it('should handle different image formats', () => {
			const imageFormats = [
				'/Images/icon.png',
				'/Images/icon.jpg',
				'/Images/icon.svg',
				'/Images/icon.webp',
			]

			imageFormats.forEach(icon => {
				const props = createMockCardProps({ icon })
				const { unmount } = render(<CardHome {...props} />)

				const image = screen.getByRole('img')
				// Next.js Image transforms URLs, check it contains the original path
				expect(image.getAttribute('src')).toContain(
					icon.replace('/Images/', '').split('.')[0],
				)
				unmount()
			})
		})
	})
})

describe('CardHome Styles', () => {
	describe('Desktop Styles', () => {
		it('should have correct global styles for desktop', () => {
			const desktopStyles = styles(false)

			expect(desktopStyles.global).toEqual({
				width: '28%',
				marginBottom: undefined,
				minWidth: '250px',
			})
		})

		it('should have correct card styles for desktop', () => {
			const desktopStyles = styles(false)

			expect(desktopStyles.card).toEqual({
				borderRadius: '10px',
				boxShadow: '1px 2px 2px 2px rgba(0, 0, 0, 0.15)',
				textAlign: 'center',
				width: '100%',
				paddingBottom: '2vh',
				height: '47vh',
				cursor: 'pointer',
				paddingTop: '5vh',
			})
		})

		it('should have correct subtitle styles for desktop', () => {
			const desktopStyles = styles(false)

			expect(desktopStyles.subtitle).toEqual({
				color: 'var(--grey)',
				fontSize: '1.8rem',
				marginBottom: '4vh',
				marginLeft: '2%',
				marginRight: '2%',
			})
		})

		it('should have correct arrow styles for desktop', () => {
			const desktopStyles = styles(false)

			expect(desktopStyles.arrow).toEqual({
				width: '1.3vw',
			})
		})
	})

	describe('Mobile Styles', () => {
		it('should have correct global styles for mobile', () => {
			const mobileStyles = styles(true)

			expect(mobileStyles.global).toEqual({
				width: '100%',
				marginBottom: '5vh',
				minWidth: undefined,
			})
		})

		it('should have correct card styles for mobile', () => {
			const mobileStyles = styles(true)

			expect(mobileStyles.card).toEqual({
				borderRadius: '10px',
				boxShadow: '1px 2px 2px 2px rgba(0, 0, 0, 0.15)',
				textAlign: 'center',
				width: '100%',
				paddingBottom: '2vh',
				height: undefined,
				cursor: 'pointer',
				paddingTop: '5vh',
			})
		})

		it('should have correct subtitle styles for mobile', () => {
			const mobileStyles = styles(true)

			expect(mobileStyles.subtitle).toEqual({
				color: 'var(--grey)',
				fontSize: '1.5rem',
				marginBottom: '2vh',
				marginLeft: '2%',
				marginRight: '2%',
			})
		})

		it('should have correct arrow styles for mobile', () => {
			const mobileStyles = styles(true)

			expect(mobileStyles.arrow).toEqual({
				width: '5vw',
			})
		})
	})

	describe('Common Styles', () => {
		it('should have consistent button styles across devices', () => {
			const desktopStyles = styles(false)
			const mobileStyles = styles(true)

			const expectedButtonStyles = {
				width: '40px',
				height: '40px',
				display: 'flex',
				alignItems: 'center',
				justifyContent: 'center',
				padding: '20px',
				marginLeft: 'auto',
				marginRight: 'auto',
				marginTop: '2vh',
				marginBottom: '4vh',
				backgroundColor: 'var(--main)',
				borderRadius: '50%',
				color: 'white',
			}

			expect(desktopStyles.button).toEqual(expectedButtonStyles)
			expect(mobileStyles.button).toEqual(expectedButtonStyles)
		})

		it('should have consistent image styles across devices', () => {
			const desktopStyles = styles(false)
			const mobileStyles = styles(true)

			const expectedImageStyles = {
				width: '20%',
				height: '20%',
				marginLeft: 'auto',
				marginRight: 'auto',
			}

			expect(desktopStyles.image).toEqual(expectedImageStyles)
			expect(mobileStyles.image).toEqual(expectedImageStyles)
		})

		it('should have consistent title heading styles across devices', () => {
			const desktopStyles = styles(false)
			const mobileStyles = styles(true)

			const expectedTitleStyles = {
				marginBottom: '0',
			}

			expect(desktopStyles.titleHeading).toEqual(expectedTitleStyles)
			expect(mobileStyles.titleHeading).toEqual(expectedTitleStyles)
		})

		it('should have consistent arrow container styles across devices', () => {
			const desktopStyles = styles(false)
			const mobileStyles = styles(true)

			const expectedArrowContainerStyles = {
				display: 'flex',
				alignItems: 'center',
			}

			expect(desktopStyles.arrowContainer).toEqual(expectedArrowContainerStyles)
			expect(mobileStyles.arrowContainer).toEqual(expectedArrowContainerStyles)
		})
	})
})
