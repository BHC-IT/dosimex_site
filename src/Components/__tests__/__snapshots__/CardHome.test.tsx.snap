// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardHome Component > Rendering > should render card with all content correctly 1`] = `
<div
  data-testid="tilt-wrapper"
  scale="1.05"
  style="width: 28%; min-width: 250px;"
  tiltmaxanglex="10"
  tiltmaxangley="10"
>
  <button
    aria-label="Navigate to Software Solutions section - Advanced radiation dosimetry software for comprehensive dose management and analysis."
    style="border-radius: 10px; box-shadow: 1px 2px 2px 2px rgba(0, 0, 0, 0.15); text-align: center; width: 100%; padding-bottom: 2vh; height: 47vh; cursor: pointer; padding-top: 5vh; background: white;"
  >
    <div
      style="width: 20%; height: 20%; margin-left: auto; margin-right: auto;"
    >
      <img
        alt="Feature icon - Software Solutions"
        data-nimg="1"
        decoding="async"
        height="80"
        loading="lazy"
        src="/_next/image?url=%2FImages%2Fsoftware-icon.png&w=256&q=80"
        srcset="/_next/image?url=%2FImages%2Fsoftware-icon.png&w=96&q=80 1x, /_next/image?url=%2FImages%2Fsoftware-icon.png&w=256&q=80 2x"
        style="color: transparent;"
        width="80"
      />
    </div>
    <h4
      style="margin-bottom: 0px;"
    >
      Software Solutions
    </h4>
    <p
      style="color: var(--grey); font-size: 1.8rem; margin-bottom: 4vh; margin-left: 2%; margin-right: 2%;"
    >
      Advanced radiation dosimetry software for comprehensive dose management and analysis.
    </p>
    <div
      style="width: 40px; height: 40px; display: flex; align-items: center; justify-content: center; padding: 20px; margin: 2vh auto 4vh auto; background-color: var(--main); border-radius: 50%; color: white;"
    >
      <div
        style="display: flex; align-items: center;"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-right-long "
          data-icon="right-long"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="width: 1.3vw;"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M334.5 414c8.8 3.8 19 2 26-4.6l144-136c4.8-4.5 7.5-10.8 7.5-17.4s-2.7-12.9-7.5-17.4l-144-136c-7-6.6-17.2-8.4-26-4.6s-14.5 12.5-14.5 22l0 72L32 192c-17.7 0-32 14.3-32 32l0 64c0 17.7 14.3 32 32 32l288 0 0 72c0 9.6 5.7 18.2 14.5 22z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </button>
</div>
`;
